go-mode.el

```diff
@@ -1562,6 +1568,185 @@ for."
       (if (not (eq cur-buffer (current-buffer)))
           (display-buffer (current-buffer) `(,go-coverage-display-buffer-func))))))
 
+(defun go-goto-function ()
+  "Go to the function defintion above point.
+
+If we are on a docstring, follow the docstring down.
+If no function is found, assume that we are at the top of a file
+and search forward instead.
+
+If we are inside an anonymous function, go to that.
+If we are inside a function that has an anonymous function inside of it and we
+are below that anonymous function, go to the root function."
+  (interactive)
+  (let ((p (point)))
+    (cond
+     ((save-excursion
+        (beginning-of-line)
+        (looking-at "^//"))
+      ;; In case we are looking at the docstring, move on forward until we are
+      ;; not anymore
+      (beginning-of-line)
+      (while (looking-at "^//")
+        (forward-line 1))
+      ;; If we are still not looking at a function, retry by calling self again.
+      (when (not (looking-at "^func"))
+        (go-goto-function)))
+
+     ((not (looking-at "^func"))
+      ;; If we are not looking at the beginning of a function line, do a regexp
+      ;; search backwards
+      (re-search-backward "\\<func\\>" nil t)
+
+      ;; If nothing is found, assume that we are at the top of the file and
+      ;; should search forward instead.
+      (when (not (looking-at "func"))
+        (re-search-forward "\\<func\\>" nil t)
+        (forward-word -1))
+
+      ;; If we have landed at an anonymous function, it is possible that we
+      ;; were not inside it but below it. If we were not inside it, we should
+      ;; go to the containing function.
+      (while (go--in-function-p p)
```

Comment:
Internally that function will call `go-goto-function` in every iteration. If point is at `<point>func`, `re-search-backwards` will search backwards further, so for every time `go--in-function-p` is `t` the loop will continue upwards until the end of the cursor is past the comparing point `p`.
The first check in `go--in-function-p` (`(looking-at "^func")`) will make it stop if it reaches a top level function (which I think only happens when you are outside a top-level function).

It's basically the algorithm you described, but wrapped in a `while` loop.

I'd agree it is a tad complicated, but I couldn't figure out a simpler way of doing it. Do you want me to try to add a better explanation in the comments?